cpp/src/io/parquet/experimental/page_index_filter.cu

```diff
@@ -0,0 +1,779 @@
+      // Only comparable types except fixed point are supported.
+      if (cudf::is_compound(dtype) && dtype.id() != cudf::type_id::STRING) {
+        // placeholder only for unsupported types.
+        columns.push_back(cudf::make_numeric_column(
+          data_type{cudf::type_id::BOOL8}, total_rows, rmm::device_buffer{}, 0, stream, mr));
+        columns.push_back(cudf::make_numeric_column(
+          data_type{cudf::type_id::BOOL8}, total_rows, rmm::device_buffer{}, 0, stream, mr));
+        return;
+      }
+      auto [min_col, max_col] = cudf::type_dispatcher<dispatch_storage_type>(
+        dtype, stats_col, schema_idx, dtype, stream, mr);
+      columns.push_back(std::move(min_col));
+      columns.push_back(std::move(max_col));
+    });
+
+  auto stats_table = cudf::table(std::move(columns));
+
+  // Converts AST to StatsAST with reference to min, max columns in above `stats_table`.
+  parquet::detail::stats_expression_converter const stats_expr{
+    filter.get(), static_cast<size_type>(output_dtypes.size())};
+
+  // Filter the input table using AST expression and return the (BOOL8) predicate column.
+  return cudf::detail::compute_column(stats_table, stats_expr.get_stats_expr().get(), stream, mr);
+}
+
+std::vector<thrust::host_vector<bool>> aggregate_reader_metadata::compute_data_page_mask(
+  cudf::column_view row_mask,
+  cudf::host_span<std::vector<size_type> const> row_group_indices,
+  cudf::host_span<cudf::data_type const> output_dtypes,
+  cudf::host_span<cudf::size_type const> output_column_schemas,
+  rmm::cuda_stream_view stream) const
+{
+  CUDF_EXPECTS(row_mask.type().id() == cudf::type_id::BOOL8,
+               "Input row bitmask should be of type BOOL8");
+
+  auto const total_rows  = row_mask.size();
+  auto const num_columns = output_dtypes.size();
+
+  auto const has_page_index =
+    compute_has_page_index(per_file_metadata, row_group_indices, output_column_schemas);
+  CUDF_EXPECTS(has_page_index,
+               "Data page mask computation requires the Parquet page index for all output columns");
+
+  // Compute page row counts, offsets, and column chunk page offsets for each column
+  std::vector<cudf::detail::host_vector<size_type>> page_row_counts;
+  std::vector<cudf::detail::host_vector<size_type>> page_row_offsets;
+  std::vector<cudf::detail::host_vector<size_type>> col_chunk_page_offsets;
+  page_row_counts.reserve(num_columns);
+  page_row_offsets.reserve(num_columns);
+  col_chunk_page_offsets.reserve(num_columns);
+
+  auto zip_iter =
+    thrust::make_zip_iterator(thrust::make_tuple(thrust::counting_iterator<cudf::size_type>(0),
+                                                 page_row_counts.begin(),
+                                                 page_row_offsets.begin(),
+                                                 col_chunk_page_offsets.begin()));
+  std::for_each(zip_iter, zip_iter + num_columns, [&](auto const& item) {
+    auto const col_idx    = thrust::get<0>(item);
```

Comment:
why is the zip iterator needed now?

Reply:
Remnant from our discussion. Thanks for catching this. Removed